From: Lennart Poettering <lennart@poettering.net>
Date: Sat, 26 Jan 2019 12:18:16 +0100
Subject: nspawn: ignore SIGPIPE for nspawn itself

Let's not abort due to a dead stdout.

Fixes: #11533
(cherry picked from commit 2949ff26911b165d3c5452df7b09471d289f9ad9)
(cherry picked from commit 97377e7f2fac37245662fdbabc29a500c6c4bca7)
---
 src/nspawn/nspawn.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/src/nspawn/nspawn.c b/src/nspawn/nspawn.c
index 91c97b6..4c42b55 100644
--- a/src/nspawn/nspawn.c
+++ b/src/nspawn/nspawn.c
@@ -4230,6 +4230,11 @@ int main(int argc, char *argv[]) {
         if (r < 0)
                 goto finish;
 
+        /* Ignore SIGPIPE here, because we use splice() on the ptyfwd stuff and that will generate SIGPIPE if
+         * the result is closed. Note that the container payload child will reset signal mask+handler anyway,
+         * so just turning this off here means we only turn it off in nspawn itself, not any children. */
+        (void) ignore_signals(SIGPIPE, -1);
+
         n_fd_passed = sd_listen_fds(false);
         if (n_fd_passed > 0) {
                 r = fdset_new_listen_fds(&fds, false);
